src/shell/mysh.c

```diff
@@ -9,13 +12,276 @@ typedef struct command {
+    int max_len = 1024;
+    char cmd_str[max_len];
+
+    command * cmd_list;
+    while(1) {
+        print_prompt();
+
+        fgets(cmd_str, max_len, stdin);
+
+        cmd_list = structure_cmds(cmd_str);
+        print_cmd_list(cmd_list);
+    }
+
+
     return 0;
 }
+
+void print_prompt() {
+    char cwd[1024];
+    getcwd(cwd, sizeof(cwd));
+    printf("%s:%s>", getlogin(), cwd);
+}
+
+command * structure_cmds(char * cmd_str) {
+    char *begin = cmd_str;
+    char *iter = cmd_str;
+    /* possible states are adding a command (0), 
+     * redirecting input (1), and redirecting output (2)
+     */
+    char state = 0;
+    char seen_char = 0;
+    int token_counter = 0;
+    command *head = NULL;
+    command *tail = NULL;
+
+    while (iter - cmd_str < strlen(cmd_str)) {
+
+        /* if double quote, skip to next double quote */
+        if (*iter == '"') {
+            iter++;
+            /* if no closing quote appears, exit */
+            while (*iter != '"') {
+                if (*iter == '\0') {
+                    fprintf(stderr, "Mismatched quotes\n");
+                    return NULL;
+                }
+                iter++;
+            }
+            seen_char = 1;
+        }
+
+        /* if whitespace */
+        else if (*iter == ' ' || *iter == '\t') {
+            /* if non-whitespace has passed, count a token */
+            if (seen_char) {
+                token_counter++;
+                seen_char = 0;
+            }
+            /* otherwise nothing (in particular, do not set seen_char to 1) */
+        }
+
+        /* check for pipes and redirects and end of string */
+        else if (*iter == '|' || *iter == '<' || *iter == '>' || 
+                 *iter == '\0' || *iter == '\n') {
+
+            /* end of one state */
+            if (seen_char) {
+                token_counter++;
+            }
+
+            /* new command */
+            if (state == 0) {
+
+                if (token_counter == 0) {
+                    fprintf(stderr, "Can't pipe to empty command\n");
+                    return 0;
+                }
+                /* create new command and set to tail */
+                if (head == NULL) {
+                    head = tail = new_command();
+                }
+                else {
+                    tail->next = new_command();
+                    tail = tail->next;
+                }
+
+                /* set tokens array */
+                tail->tokens = 
+                    (char **) malloc(sizeof(char *) * (token_counter + 1));
+                split_cmd(tail->tokens, begin, iter, token_counter);
+                /* set final element of tokens array to NULL */
+                tail->tokens[token_counter] = NULL;
+            }
+
+            /* redirect input */
+            else if (state == 1) {
+                if (token_counter != 1) {
+                    fprintf(stderr, "Input must be one token\n");
+                    return 0;
```

Comment:
Could this leak memory? Maybe we need to call cleanup_commands(head) or something before returning.

Nit on style: Let's return NULL instead of return 0?